### ISSUE TITLE
JavaParserTypeSolver is not solving symbols

### ISSUE THREAD

**Issue by wozuo** [javaparser/javaparser#1872]
Hey everyone,

Sorry to bombard you with the next issue, but I'm trying to add a JavaParserTypeSolver to a combinedTypeSolver and it doesn't find the classes it should. This is what I'm doing:

First I create a new `CombinedTypeSolver` and add some type solvers, create a parserConfiguration and then I collect all source roots in `projectFolder.toPath()`:
```
CombinedTypeSolver combinedTypeSolver = new CombinedTypeSolver();
combinedTypeSolver.add(new JavaParserTypeSolver(new File("/Users/marc/Downloads/okhttp-parent-3.8.0/okhttp/src/main/java")));
combinedTypeSolver.add(new ReflectionTypeSolver());
ParserConfiguration parserConfiguration = new ParserConfiguration().setSymbolResolver(new JavaSymbolSolver(combinedTypeSolver));
projectRoot = new SymbolSolverCollectionStrategy(parserConfiguration).collect(projectFolder.toPath());
```

After that I'm looping through all the source roots of the project:
```
for (SourceRoot sourceRoot : projectRoot.getSourceRoots()) {
    try {
        sourceRoot.tryToParse();
        compilationUnits = sourceRoot.getCompilationUnits();
    } catch (IOException e) {
        e.printStackTrace();
        return;
    }
}
```

For each of the compilation units I then look for example for `ObjectCreationExpr` expressions and would like to resolve their types (I will leave out the unimportant code here and just paste the code I'm using to resolve the expression):
```
((ObjectCreationExpr) expr).resolve().getClassName();
``` 

The source code I'm parsing imports classes from the library okhttp like so:
```
import okhttp3.OkHttpClient;
```
And an example for such an object creation I mentioned above is:
```
mHttpClient = new OkHttpClient.Builder();
```

Now if I understand the JavaParserTypeSolver correctly, it should look for `/Users/marc/Downloads/okhttp-parent-3.8.0/okhttp/src/main/java/okhttp3/OkHttpClient.java` which exists on my filesystem but it doesn't resolve the types correctly and instead throws an `UnsolvedSymbolException`.

Am I doing anything wrong or is there a bug in the type solver?

Thanks and best regards,
Marc

**Comment by malteskoruppa** [comment]
To me it looks like you're doing everything correctly.

This is just a guess and @ftomassetti will be able to make a better judgment, but I think this may be a missing functionality in the symbol solver. The thing is that the `Builder` class that the code instantiates is a public static member class within the class `OkHttpClient`, and the symbol solver may not handle this case properly.

**Comment by wozuo** [comment]
Hey Malte,

Thanks for your answers and effort! I just can't shake the feeling that I may be doing something wrong, because today I tested more code (a few simpler expressions) and one example is the following code that can't be resolved:
The import is `import org.apache.http.impl.client.DefaultHttpClient;` and the code snippet I'm trying to resolve is the ObjectCreationExpr in `HttpClient client = new DefaultHttpClient();` which throws an `UnsolvedSymbolException`. And I'm adding the JavaParserTypeSolver: `combinedTypeSolver.add(new JavaParserTypeSolver(new File("/Users/marc/Downloads/httpcomponents-client-4.0/httpclient/src/main/java")));`.

best regards,
Marc

**Comment by malteskoruppa** [comment]
Hey Marc,

I invested some more time and I agree, I think you're doing something wrong. But I'm not sure what. However, I downloaded the OkHttp package mentioned in your original code, and I looked for an object creation expression `new OkHttpClient.Builder()` in there, and I tried to resolve it, and it worked. So (probably) not a bug in the symbol solver. :-)

Here's what I did in detail. Can you reproduce it?

1. I downloaded OkHttp 3.8.0 from here: https://github.com/square/okhttp/archive/parent-3.8.0.tar.gz

2. I unpacked the archive inside the folder `/home/malte/Downloads`.

3. I wrote the following test method, and it passes:

```java
    @Test
    public void foo() throws FileNotFoundException {
        // configure symbol solver
        CombinedTypeSolver combinedTypeSolver = new CombinedTypeSolver();
        combinedTypeSolver.add(new ReflectionTypeSolver());
        combinedTypeSolver.add(new JavaParserTypeSolver("/home/malte/Downloads/okhttp-parent-3.8.0/okhttp/src/main/java"));
        JavaParser.getStaticConfiguration().setSymbolResolver(new JavaSymbolSolver(combinedTypeSolver));

        // get the following object creation expression:
        //     OkHttpClient.Builder builder = new OkHttpClient.Builder();
        // from okhttp3/curl/Main.java, class Main, method createClient()
        CompilationUnit cu = JavaParser.parse(new File("/home/malte/Downloads/okhttp-parent-3.8.0/okcurl/src/main/java/okhttp3/curl/Main.java"));
        ClassOrInterfaceDeclaration clazz = cu.getClassByName("Main").get();
        MethodDeclaration method = clazz.getMethodsByName("createClient").get(0);
        ObjectCreationExpr expr = method.findFirst(ObjectCreationExpr.class).get();

        // just make sure we got the right object creation expression
        assertEquals("new OkHttpClient.Builder()", expr.toString());

        // resolve method call expression
        ResolvedConstructorDeclaration declaration = expr.resolve(); // works fine

        assertEquals("okhttp3.OkHttpClient.Builder.Builder()", declaration.getQualifiedSignature());
    }
```

**Comment by wozuo** [comment]
Hey thanks I just tested it and this works for me too. This is really weird...

So maybe the problem is the way I add the configuration to `SymbolSolverCollectionStrategy` instead of doing ` JavaParser.getStaticConfiguration().setSymbolResolver(...)`?

**Comment by malteskoruppa** [comment]
After experimenting some more, I think I know what's going wrong. The problem is that the combined type solver that you initialize gets overwritten by the collection strategy. The collection strategy builds its own combined type solver and just ignores the one that you configured previously.

You do this:
 ```java
ParserConfiguration parserConfiguration = new ParserConfiguration().setSymbolResolver(new JavaSymbolSolver(combinedTypeSolver));
projectRoot = new SymbolSolverCollectionStrategy(parserConfiguration).collect(projectFolder.toPath());
```

That is, you create a parser configuration and you set this configuration to use your correctly configured combined type solver, which includes the `JavaParserSymbolSolver` for the okhttp project.
Then, you instantiate a new `SymbolSolverCollectionStrategy` and pass it your `parserConfiguration`. And what the constructor of `SymbolSolverCollectionStrategy`  does is just this:
```java
this.parserConfiguration = parserConfiguration.setSymbolResolver(new JavaSymbolSolver(typeSolver));
```

In other words, it overwrites your parser configuration's symbol resolver with its own type solver (which is initially a combined type solver with only a `ReflectionTypeSolver`, and then, the collection strategy apparently adds `JavaParserSymbolSolver`s for each source root that it finds to this combined type solver).

However, the `JavaParserSymbolSolver` for that `okhttp` project that you initialized yourself is ignored and overwritten. And since the okhttp project is presumably not among the source roots found by the collection strategy, it is not added this way either.

So that explains the unsolved symbol exception.

Unfortunately I am not familiar with the collection strategy, I haven't used it before. I don't know whether it's possible to give it some additional type solvers that it should use (from a quick look, I couldn't find a possibility to do so), or whether such scenarios are intended to be solved in some other manner altogether.

@daanschipper will be probably able to help you. :)

**Comment by matozoid** [comment]
Here's a @daanschipper in a box: http://javaparser.org/setting-up-for-analysing-a-whole-project/ :-)

**Comment by wozuo** [comment]
Ah yea then that's it! Is this the intended behavior that the symbol solver is overwritten? I mean what's the point of passing it in `SymbolSolverCollectionStrategy(parserConfiguration)` if it is overwritten afterwards? One solution would be to just do `this.parserConfiguration = parserConfiguration;` in the constructor of `SymbolSolverCollectionStrategy`, then one would need to pass in the `ReflectionTypeSolver ` but that's not really a problem right?

Thanks a lot @malteskoruppa for solving this! I'll just do a workaround for now.

**Comment by malteskoruppa** [comment]
Thanks @matozoid for the link! :-)

Unfortunately, I can't find any information on how to solve this particular problem there either. :(

Marc, in the interim, you can modify the constructor of `SymbolSolverCollectionStrategy` as follows. Where it currently has this code:
```java
    private final CombinedTypeSolver typeSolver = new CombinedTypeSolver(new ReflectionTypeSolver(false));

    public SymbolSolverCollectionStrategy() {
        this(new ParserConfiguration());
    }

    public SymbolSolverCollectionStrategy(ParserConfiguration parserConfiguration) {
        this.parserConfiguration = parserConfiguration.setSymbolResolver(new JavaSymbolSolver(typeSolver));
    }
```

Change this like so:
```java
    private final CombinedTypeSolver typeSolver;

    public SymbolSolverCollectionStrategy() {
        this(new ParserConfiguration());
    }

    public SymbolSolverCollectionStrategy(ParserConfiguration parserConfiguration) {
        this(parserConfiguration, new CombinedTypeSolver(new ReflectionTypeSolver(false)));
    }

    public SymbolSolverCollectionStrategy(ParserConfiguration parserConfiguration, CombinedTypeSolver typeSolver) {
        this.typeSolver = typeSolver;
        this.parserConfiguration = parserConfiguration.setSymbolResolver(new JavaSymbolSolver(this.typeSolver));
    }
```

That way, you have a constructor which uses an initial combined type solver that you can specify yourself. Then, it will work. This solution doesn't destroy any functionality. In particular, `JavaParserSymbolSolver`s for any source root folders found by the collection strategy are still added to the combined type solver that you initialized yourself. (So be aware that your combined type solver may "grow" when you pass it into the constructor of the collection strategy.)

There may be better ways, but this should do just fine until @daanschipper provides a better solution. :)

**Comment by malteskoruppa** [comment]
> what's the point of passing it in `SymbolSolverCollectionStrategy(parserConfiguration)` if it is overwritten afterwards?

I guess that the point is that you can configure everything (e.g., the Java language level, etc.) _except_ for the symbol resolver, which is initialized by the collection strategy. That's the very point of the collection strategy, if I understand it correctly. :) It just seems to lack a way of manually adding more type solvers to it, or at least I don't understand how this could currently be done without modifications to its implementation.

**Comment by malteskoruppa** [comment]
> One solution would be to just do `this.parserConfiguration = parserConfiguration;` in the constructor of `SymbolSolverCollectionStrategy`, then one would need to pass in the `ReflectionTypeSolver` but that's not really a problem right?

That's not enough. Then the collection strategy won't add `JavaParserSymbolSolver`s for the source roots it finds to your symbol resolver, but to its own (unused) combined type solver. That's likely not what you want. The solution that I provided above takes this into account.

Sure, you need to pass in your own `ReflectionTypeSolver` into your combined type solver, but that's not a problem indeed. Actually, this way you even have more flexibility, since you can also instantiate it with `jreOnly` set to `true` (the collection strategy always uses `false`). I mean, I'm not really sure why one _would_ want to set that to `true` (I personally always set it to `false`), but maybe there are use cases for this. :-)

**Comment by matozoid** [comment]
The `SymbolSolverCollectionStrategy` is a utility that sets up your type solvers for you (if you have a simple project setup.) That's it really.

**Comment by malteskoruppa** [comment]
As far as I understand it, the OP basically has a project folder (`projectFolder`, with several source root folders) _and_ some external source root folder (`/Users/marc/Downloads/okhttp-parent-3.8.0/okhttp/src/main/java`) and he wants to configure the symbol solver to add type solvers for all source root folders in that project folder _and_ a type solver for the external source root folder. What would be the best way to do that?

Would it perhaps be enough to call `collect` twice on the _same_ `SymbolSolverCollectionStrategy`, once with the `projectFolder` and once with the `okhttp-parent-3.8.0` folder?

*Edit:* Oh yeah, that works. Duh! :laughing: 

@wozuo You don't have to do any modifications to the source code. You don't have to create your own combined type solver. All you have to do is this:

```java
SymbolSolverCollectionStrategy symbolSolverCollectionStrategy = new SymbolSolverCollectionStrategy();
ProjectRoot projectRoot = symbolSolverCollectionStrategy.collect(projectFolder.toPath());
// also add a type solver for the source root of okhttp3
symbolSolverCollectionStrategy.collect(Paths.get("/Users/marc/Downloads/okhttp-parent-3.8.0/okhttp/src/main/java"));
```

Assuming that you may want to add the entire `okhttp-parent` library to your type solvers, you may even make that last line simply:

```java
symbolSolverCollectionStrategy.collect(Paths.get("/Users/marc/Downloads/okhttp-parent-3.8.0"));
```

...and the collection strategy will hopefully find all source root folders there too! :-)

**Comment by wozuo** [comment]
Oh nice thats a clever solution! 

But if someone wants to use other type solvers as well (e.g. JarTypeSolver) then you would still need a way to add it to the `SymbolSolverCollectionStrategy `.

**Comment by malteskoruppa** [comment]
> Oh nice thats a clever solution!

Thanks. Actually I assume it's the intended way. ;) I just didn't realize it immediately, since I never used the collection strategy before and I wasn't familiar with it.

> But if someone wants to use other type solvers as well (e.g. JarTypeSolver) then you would still need a way to add it to the `SymbolSolverCollectionStrategy`.

Again, not an expert on the collection strategy, but you can solve this problem in exactly the same way. When you call `collect` on a directory that contains jar files somewhere (or you call `collect` on a jar file directly), type solvers for those jars are added to the collection strategy as well.

Basically, just call `collect` on all the project folders and all the jar files relevant to you (or even some directory that contains a collection of jar files). It will do everything for you. Nifty tool. :)

**Comment by fuhrmanator** [comment]
Updating the [blog entry](http://javaparser.org/setting-up-for-analysing-a-whole-project/) to show multiple `collect` calls would help a lot, since I also had to do it for similar reasons (external Jars). Indeed I was happy to find the facade, but ultimately had to spend time to finally find the final answer here.
